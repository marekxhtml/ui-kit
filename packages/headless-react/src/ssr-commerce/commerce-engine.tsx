import {
  Controller,
  CommerceEngine,
  ControllerDefinitionsMap,
  CommerceEngineDefinitionOptions,
  defineCommerceEngine as defineBaseCommerceEngine,
  CommerceEngineOptions,
  SolutionType, // getSampleCommerceEngineConfiguration,
  // defineSummary,
  // defineProductList,
  // defineRecommendations,
  // defineCart,
  // defineSearchBox,
  // defineContext,
  // defineRecentQueriesList,
  // defineNotifyTrigger,
  // defineQueryTrigger,
  // defineRedirectionTrigger,
  // defineStandaloneSearchBox,
  // defineInstantProducts,
  // definePagination,
  // defineSort,
  // defineProductView,
  // defineDidYouMean,
  // defineFacetGenerator,
} from '@coveo/headless/ssr-commerce';
// Workaround to prevent Next.js erroring about importing CSR only hooks
import React from 'react';
import {singleton, SingletonGetter} from '../utils.js';
import {
  buildControllerHooks,
  buildEngineHook,
  buildHydratedStateSearchProvider,
  buildHydratedStateStandaloneProvider,
  buildStaticStateListingProvider,
  buildStaticStateSearchProvider,
  buildStaticStateStandaloneProvider,
} from './common.js';
import {ContextState, ReactEngineDefinition} from './types.js';

export type ReactCommerceEngineDefinition<
  TControllers extends ControllerDefinitionsMap<CommerceEngine, Controller>,
  TSolutionType extends SolutionType,
> = ReactEngineDefinition<
  CommerceEngine,
  TControllers,
  CommerceEngineOptions,
  TSolutionType
>;

// Wrapper to workaround the limitation that `createContext()` cannot be called directly during SSR in next.js
export function createSingletonContext<
  TControllers extends ControllerDefinitionsMap<CommerceEngine, Controller>,
  TSolutionType extends SolutionType = SolutionType,
>() {
  return singleton(() =>
    React.createContext<ContextState<
      CommerceEngine,
      TControllers,
      TSolutionType
    > | null>(null)
  );
}

/**
 * Returns controller hooks as well as SSR and CSR context providers that can be used to interact with a Commerce engine
 *  on the server and client side respectively.
 */
export function defineCommerceEngine<
  TControllers extends ControllerDefinitionsMap<CommerceEngine, Controller>,
>(options: CommerceEngineDefinitionOptions<TControllers>) {
  const singletonContext = createSingletonContext<TControllers>();

  type ContextStateType<TSolutionType extends SolutionType> = SingletonGetter<
    React.Context<ContextState<
      CommerceEngine,
      TControllers,
      TSolutionType
    > | null>
  >;
  type ListingContext = ContextStateType<SolutionType.listing>;
  type SearchContext = ContextStateType<SolutionType.search>;
  type StandaloneContext = ContextStateType<SolutionType.standalone>;

  const {
    listingEngineDefinition,
    searchEngineDefinition,
    standaloneEngineDefinition,
  } = defineBaseCommerceEngine({...options});
  return {
    useEngine: buildEngineHook(singletonContext),
    controllers: buildControllerHooks(singletonContext, options.controllers),
    listingEngineDefinition: {
      ...listingEngineDefinition,
      StaticStateProvider: buildStaticStateListingProvider(
        singletonContext as ListingContext
      ),

      HydratedStateProvider: buildStaticStateListingProvider(
        singletonContext as ListingContext
      ),
    },
    searchEngineDefinition: {
      ...searchEngineDefinition,
      StaticStateProvider: buildStaticStateSearchProvider(
        singletonContext as SearchContext
      ),
      HydratedStateProvider: buildHydratedStateSearchProvider(
        singletonContext as SearchContext
      ),
    },
    standaloneEngineDefinition: {
      ...standaloneEngineDefinition,
      StaticStateProvider: buildStaticStateStandaloneProvider(
        singletonContext as StandaloneContext
      ),
      HydratedStateProvider: buildHydratedStateStandaloneProvider(
        singletonContext as StandaloneContext
      ),
    },
  };
}
