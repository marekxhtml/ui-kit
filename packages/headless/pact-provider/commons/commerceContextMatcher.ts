import {MatchersV3} from '@pact-foundation/pact';
import {SortBy} from '../../src/features/sort/sort.js';

const {
  equal,
  regex,
  like,
  string,
  integer,
  boolean,
  atLeastOneLike,
  eachLike,
  eachKeyMatches,
} = MatchersV3;

export const commerceContextMatcher = like({
  user: like({
    userAgent: string(
      'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (HTML, like Gecko) Chrome/115.0.0.0 Safari/537.36'
    ),
    // latitude: decimal(46.8131),
    // longitude: decimal(71.2075),
  }),
  view: eachKeyMatches(
    {
      url: string(),
      referrer: string(),
    },
    regex('url|referrer', 'url')
  ),
  cart: eachLike(
    {
      productId: string('shoe-a1-red'),
      quantity: integer(2),
    },
    0
  ),
  source: atLeastOneLike(string('@coveo/headless@2.61.0')),
  capture: boolean(true),
  // labels: like({
  //   category: string('garden > garden-tools > chainsaws'),
  //   brand: string('ACME'),
  // }),
});

export const commercePagerMatcher = {perPage: integer(30)};

export const sortByRelevanceMatcher = {sortCriteria: equal(SortBy.Relevance)};
export const sortByFieldsMatcher = {
  sortCriteria: equal(SortBy.Fields),
  fields: eachLike({
    field: string('price'),
    ...eachKeyMatches(
      {
        direction: regex('asc|desc', 'asc'),
      },
      regex('field|direction', 'asc')
    ),
  }),
};
