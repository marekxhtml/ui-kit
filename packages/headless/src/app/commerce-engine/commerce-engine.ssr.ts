/* eslint-disable @typescript-eslint/no-unused-vars */

/**
 * Utility functions to be used for Commerce Server Side Rendering.
 */
import {Action, AnyAction, Middleware, UnknownAction} from '@reduxjs/toolkit';
import {stateKey} from '../../app/state-key';
import {buildProductListing} from '../../controllers/commerce/product-listing/headless-product-listing';
import {buildSearch} from '../../controllers/commerce/search/headless-search';
import type {Controller} from '../../controllers/controller/headless-controller';
import {
  createWaitForActionMiddleware,
  createWaitForActionMiddlewareWithCount,
} from '../../utils/utils';
import {buildControllerDefinitions} from '../commerce-ssr-engine/common';
import {
  ControllerDefinitionsMap,
  InferControllerStaticStateMapFromDefinitionsWithSolutionType,
  SolutionType,
} from '../commerce-ssr-engine/types/common';
import {
  EngineDefinition,
  EngineDefinitionOptions,
} from '../commerce-ssr-engine/types/core-engine';
import {NavigatorContextProvider} from '../navigatorContextProvider';
import {composeFunction} from '../ssr-engine/common';
import {createStaticState} from '../ssr-engine/common';
import {
  EngineStaticState,
  InferControllerPropsMapFromDefinitions,
} from '../ssr-engine/types/common';
import {
  CommerceEngine,
  CommerceEngineOptions,
  buildCommerceEngine,
} from './commerce-engine';

/**
 * The SSR commerce engine.
 */
export interface SSRCommerceEngine extends CommerceEngine {
  /**
   * Waits for the search to be completed and returns a promise that resolves to a `SearchCompletedAction`.
   */
  // TODO: document: essentially a promise for the first search/request and a list of recommendation promises
  waitForRequestCompletedActions(): Promise<Action[]>;
}

export type CommerceEngineDefinitionOptions<
  TControllers extends ControllerDefinitionsMap<SSRCommerceEngine, Controller>,
> = EngineDefinitionOptions<CommerceEngineOptions, TControllers>;

function isListingFetchCompletedAction(action: unknown): action is Action {
  return /^commerce\/productListing\/fetch\/(fulfilled|rejected)$/.test(
    (action as UnknownAction).type
  );
}

function isSearchCompletedAction(action: unknown): action is Action {
  return /^commerce\/search\/executeSearch\/(fulfilled|rejected)$/.test(
    (action as UnknownAction).type
  );
}

function isRecommendationCompletedAction(action: unknown): action is Action {
  return /^commerce\/recommendations\/fetch\/(fulfilled|rejected)$/.test(
    (action as UnknownAction).type
  );
}

// TODO: rename
function getRecommendationControllerCountFromDefinitions<
  TControllerDefinitions extends ControllerDefinitionsMap<
    SSRCommerceEngine,
    Controller
  >,
>(definitionMap: TControllerDefinitions): number {
  let slotIdCount = 0;
  // TODO: remove any
  Object.values(definitionMap as Record<string, any>).forEach((x) => {
    if (x.isRecommendation) {
      slotIdCount++;
    }
  });
  return slotIdCount;
}

// TODO: rename
function getRecommendationWaitForActionMiddleware<
  TControllerDefinitions extends ControllerDefinitionsMap<
    SSRCommerceEngine,
    Controller
  >,
>(
  definitionMap: TControllerDefinitions
): {
  promise: Promise<Action>;
  middleware: Middleware;
} {
  let slotIdCount =
    getRecommendationControllerCountFromDefinitions(definitionMap);

  return createWaitForActionMiddlewareWithCount(
    isRecommendationCompletedAction,
    slotIdCount
  );
}

function buildSSRCommerceEngine(
  solutionType: SolutionType,
  options: CommerceEngineOptions & {
    recommendationOptions: {
      // TODO: rename
      promise: Promise<Action>;
      middleware: Middleware;
    };
  }
): SSRCommerceEngine {
  // TODO: put in function

  // let waitAction;
  // if (solutionType === SolutionType.recommendation) {
  //   // in case of recommendation only (e.g. home page)
  //   // TODO: clean that
  //   const middleware: Middleware = () => (next) => (action) => {
  //     next(action);
  //   };
  //   waitAction = {middleware, promise: Promise.resolve(null as unknown as Action)}; // TODO: get rid of that
  // } else {
  if (solutionType !== SolutionType.recommendation) {
    const {middleware, promise} = createWaitForActionMiddleware(
      solutionType === SolutionType.listing
        ? isListingFetchCompletedAction
        : isSearchCompletedAction
    );
    const commerceEngine = buildCommerceEngine({
      ...options,
      middlewares: [
        ...(options.middlewares ?? []),
        middleware,
        options.recommendationOptions.middleware, // TODO: clean that
      ],
    });
    return {
      ...commerceEngine,

      get [stateKey]() {
        return commerceEngine[stateKey];
      },

      waitForRequestCompletedActions() {
        return Promise.all([promise, options.recommendationOptions.promise]); // TODO: clean that
      },
    };
  } else {
    const commerceEngine = buildCommerceEngine({
      ...options,
      middlewares: [
        ...(options.middlewares ?? []),
        options.recommendationOptions.middleware, // TODO: clean that
      ],
    });
    return {
      ...commerceEngine,

      get [stateKey]() {
        return commerceEngine[stateKey];
      },

      waitForRequestCompletedActions() {
        return Promise.all([options.recommendationOptions.promise]); // TODO: clean that
      },
    };
  }
}

export interface CommerceEngineDefinition<
  TControllers extends ControllerDefinitionsMap<SSRCommerceEngine, Controller>,
  TSolutionType extends SolutionType,
> extends EngineDefinition<
    SSRCommerceEngine,
    TControllers,
    CommerceEngineOptions,
    TSolutionType
  > {}

/**
 * Initializes a Commerce engine definition in SSR with given controllers definitions and commerce engine config.
 * @param options - The commerce engine definition
 * @returns Three utility functions to fetch the initial state of the engine in SSR, hydrate the state in CSR,
 *  and a build function that can be used for edge cases requiring more control.
 */
export function defineCommerceEngine<
  TControllerDefinitions extends ControllerDefinitionsMap<
    SSRCommerceEngine,
    Controller
  >,
>(
  options: CommerceEngineDefinitionOptions<TControllerDefinitions>
): {
  listingEngineDefinition: CommerceEngineDefinition<
    TControllerDefinitions,
    SolutionType.listing
  >;
  searchEngineDefinition: CommerceEngineDefinition<
    TControllerDefinitions,
    SolutionType.search
  >;
  recommendationEngineDefinition: CommerceEngineDefinition<
    TControllerDefinitions,
    SolutionType.recommendation
  >;
} {
  const {controllers: controllerDefinitions, ...engineOptions} = options;
  type Definition = CommerceEngineDefinition<
    TControllerDefinitions,
    SolutionType
  >;
  type BuildFunction = Definition['build'];
  type FetchStaticStateFunction = Definition['fetchStaticState'];
  type HydrateStaticStateFunction = Definition['hydrateStaticState'];
  type FetchStaticStateFromBuildResultFunction =
    FetchStaticStateFunction['fromBuildResult'];
  type HydrateStaticStateFromBuildResultFunction =
    HydrateStaticStateFunction['fromBuildResult'];
  type BuildParameters = Parameters<BuildFunction>;
  type FetchStaticStateParameters = Parameters<FetchStaticStateFunction>;
  type HydrateStaticStateParameters = Parameters<HydrateStaticStateFunction>;
  type FetchStaticStateFromBuildResultParameters =
    Parameters<FetchStaticStateFromBuildResultFunction>;
  type HydrateStaticStateFromBuildResultParameters =
    Parameters<HydrateStaticStateFromBuildResultFunction>;

  const getOptions = () => {
    return engineOptions;
  };

  const setNavigatorContextProvider = (
    navigatorContextProvider: NavigatorContextProvider
  ) => {
    engineOptions.navigatorContextProvider = navigatorContextProvider;
  };

  const buildFactory =
    <T extends SolutionType>(solutionType: T) =>
    async (...[buildOptions]: BuildParameters) => {
      // TODO: rename
      const recommendationOptions = getRecommendationWaitForActionMiddleware(
        controllerDefinitions ?? {}
      );
      const engine = buildSSRCommerceEngine(solutionType, {
        recommendationOptions,
        ...(buildOptions?.extend
          ? await buildOptions.extend(getOptions())
          : getOptions()),
      });
      const controllers = buildControllerDefinitions({
        definitionsMap: (controllerDefinitions ?? {}) as TControllerDefinitions,
        engine,
        solutionType,
        propsMap: (buildOptions && 'controllers' in buildOptions
          ? buildOptions.controllers
          : {}) as InferControllerPropsMapFromDefinitions<TControllerDefinitions>,
      });

      return {
        engine,
        controllers,
      };
    };

  const fetchStaticStateFactory: (
    solutionType: SolutionType
  ) => FetchStaticStateFunction = (solutionType: SolutionType) =>
    composeFunction(
      async (...params: FetchStaticStateParameters) => {
        if (!getOptions().navigatorContextProvider) {
          // TODO: KIT-3409 - implement a logger to log SSR warnings/errors
          console.warn(
            '[WARNING] Missing navigator context in server-side code. Make sure to set it with `setNavigatorContextProvider` before calling fetchStaticState()'
          );
        }
        const buildResult = await buildFactory(solutionType)(...params);
        const staticState = await fetchStaticStateFactory(
          solutionType
        ).fromBuildResult({
          buildResult,
        });
        return staticState;
      },
      {
        fromBuildResult: async (
          ...params: FetchStaticStateFromBuildResultParameters
        ) => {
          const [
            {
              buildResult: {engine, controllers},
            },
          ] = params;

          // TODO: do not execute the first request if only rendering recommendation
          if (solutionType === SolutionType.listing) {
            console.log('>> fetch listing products');
            buildProductListing(engine).executeFirstRequest();
          }

          if (solutionType === SolutionType.search) {
            console.log('>> execute search');
            buildSearch(engine).executeFirstSearch();
          }

          // Refreshing recommendation regardless of the solution type
          console.log('*********************');
          console.log(params[0].buildResult.controllers);
          console.log('*********************');

          Object.entries(controllerDefinitions as Record<string, any>).forEach(
            ([key, value]) => {
              if ((value as any).isRecommendation) {
                console.log('refreshing', key);
                key in controllers && (controllers as any)[key].refresh(); // TODO: remove any
              }
            }
          );

          return createStaticState({
            searchActions: await engine.waitForRequestCompletedActions(),
            controllers,
          }) as EngineStaticState<
            AnyAction,
            InferControllerStaticStateMapFromDefinitionsWithSolutionType<
              TControllerDefinitions,
              SolutionType
            >
          >;
        },
      }
    );

  const hydrateStaticStateFactory: (
    solutionType: SolutionType
  ) => HydrateStaticStateFunction = (solutionType: SolutionType) =>
    composeFunction(
      async (...params: HydrateStaticStateParameters) => {
        if (!getOptions().navigatorContextProvider) {
          // TODO: KIT-3409 - implement a logger to log SSR warnings/errors
          console.warn(
            '[WARNING] Missing navigator context in client-side code. Make sure to set it with `setNavigatorContextProvider` before calling hydrateStaticState()'
          );
        }
        const buildResult = await buildFactory(solutionType)(
          ...(params as BuildParameters)
        );
        const staticState = await hydrateStaticStateFactory(
          solutionType
        ).fromBuildResult({
          buildResult,
          searchActions: params[0]!.searchActions,
        });
        return staticState;
      },
      {
        fromBuildResult: async (
          ...params: HydrateStaticStateFromBuildResultParameters
        ) => {
          const [
            {
              buildResult: {engine, controllers},
              searchActions,
            },
          ] = params;

          searchActions.forEach((action) => {
            engine.dispatch(action);
          });
          await engine.waitForRequestCompletedActions();
          return {engine, controllers};
        },
      }
    );
  return {
    listingEngineDefinition: {
      build: buildFactory(SolutionType.listing),
      fetchStaticState: fetchStaticStateFactory(SolutionType.listing),
      hydrateStaticState: hydrateStaticStateFactory(SolutionType.listing),
      setNavigatorContextProvider,
    } as CommerceEngineDefinition<TControllerDefinitions, SolutionType.listing>,
    searchEngineDefinition: {
      build: buildFactory(SolutionType.search),
      fetchStaticState: fetchStaticStateFactory(SolutionType.search),
      hydrateStaticState: hydrateStaticStateFactory(SolutionType.search),
      setNavigatorContextProvider,
    } as CommerceEngineDefinition<TControllerDefinitions, SolutionType.search>,
    recommendationEngineDefinition: {
      //TODO: only used for pages with no search
      build: buildFactory(SolutionType.recommendation),
      fetchStaticState: fetchStaticStateFactory(SolutionType.recommendation),
      hydrateStaticState: hydrateStaticStateFactory(
        SolutionType.recommendation
      ),
      setNavigatorContextProvider,
    } as CommerceEngineDefinition<
      TControllerDefinitions,
      SolutionType.recommendation
    >,
  };
}
