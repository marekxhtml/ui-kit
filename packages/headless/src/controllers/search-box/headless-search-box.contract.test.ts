import {PactV3} from '@pact-foundation/pact';
import {describe, it, expect} from 'vitest';
import {sortByFieldsMatcher} from '../../../pact-provider/commons/commerceContextMatcher.js';
import {
  providerFactory,
  addSuggestionInteraction,
  addCommerceSearchInteraction,
  addCommerceSuggestionInteraction,
} from '../../../pact-provider/provider.js';
import {getSampleCommerceEngineConfiguration} from '../../app/commerce-engine/commerce-engine-configuration.js';
import {buildCommerceEngine} from '../../app/commerce-engine/commerce-engine.js';
import {
  buildSearchEngine,
  getSampleSearchEngineConfiguration,
} from '../../app/search-engine/search-engine.js';
import {
  buildFieldsSortCriterion,
  SortDirection,
} from '../../features/sort/sort.js';
import {buildSearchBox as buildCommerceSearchBox} from '../commerce/search-box/headless-search-box.js';
import {buildSearch} from '../commerce/search/headless-search.js';
import {Controller} from '../controller/headless-controller.js';
import {buildSearchBox} from './headless-search-box.js';

const getSearchBoxController = (url: string) => {
  const configuration = getSampleSearchEngineConfiguration();
  configuration.search!.proxyBaseUrl = url;
  const engine = buildSearchEngine({
    configuration,
    navigatorContextProvider: () => ({
      clientId: crypto.randomUUID(),
      location: 'some location',
      referrer: 'some referrer',
      userAgent: 'some user agent',
    }),
  });
  return buildSearchBox(engine);
};

const getCommerceControllers = (url: string) => {
  const configuration = getSampleCommerceEngineConfiguration();
  configuration.proxyBaseUrl = url;
  const engine = buildCommerceEngine({
    configuration,
    navigatorContextProvider: () => ({
      clientId: crypto.randomUUID(),
      location: 'some location',
      referrer: 'some referrer',
      userAgent: 'some user agent',
    }),
  });
  const search = buildSearch(engine);

  return {
    products: search,
    searchBox: buildCommerceSearchBox(engine),
    pagination: search.pagination(),
    sort: search.sort(),
  };
};

describe('SearchBox', () => {
  let provider: PactV3;
  beforeEach(() => {
    vi.spyOn(Math, 'random').mockReturnValue(0.5);
    provider = providerFactory();
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  describe('when there are suggestions available', () => {
    beforeEach(() => {
      provider = addSuggestionInteraction(provider);
    });

    it('should return suggestions', async () => {
      await provider.executeTest(async (mockServer) => {
        const searchBox = getSearchBoxController(mockServer.url);
        searchBox.showSuggestions();
        await waitForNSubscribeCallback(searchBox, 2);
        expect(searchBox.state).toMatchSnapshot();
      });
    });
  });

  // TODO: move this to commerce contract testing
  describe('when there are search results available', () => {
    beforeEach(() => {
      provider = addCommerceSuggestionInteraction(provider);
      provider = addCommerceSearchInteraction(provider);
    });

    it('should return search results', async () => {
      await provider.executeTest(async (mockServer) => {
        const {searchBox, products} = getCommerceControllers(mockServer.url);
        searchBox.updateText('blue shoes');
        await waitForNSubscribeCallback(searchBox, 2);
        searchBox.submit();
        await waitForNSubscribeCallback(searchBox, 2);
        expect(products.state.products).toMatchSnapshot();
      });
    });
    // TODO: test with pager and sort criteria
  });

  // TODO: put in loadmore contract testing
  describe('when sorting by fields', () => {
    beforeEach(() => {
      provider = addCommerceSearchInteraction(
        provider,
        'Sorting results by fields',
        {
          sort: sortByFieldsMatcher,
        }
      );
    });

    it('should return search results', async () => {
      await provider.executeTest(async (mockServer) => {
        const {sort, products} = getCommerceControllers(mockServer.url);
        sort.sortBy(
          buildFieldsSortCriterion([
            {name: 'custom_field', direction: SortDirection.Ascending},
          ])
        );
        await waitForNSubscribeCallback(products, 2);
        expect(sort.state).toMatchSnapshot();
      });
    });
    // TODO: test with pager and sort criteria
  });

  describe('when there are more products available', () => {
    beforeEach(() => {
      provider = addCommerceSearchInteraction(
        provider
        // 'There are search results with pagination',
        // commercePagerMatcher
      );
    });

    it('should return search results', async () => {
      await provider.executeTest(async (mockServer) => {
        const {pagination, products} = getCommerceControllers(mockServer.url);
        pagination.setPageSize(48); // Mock the pageSize returned by the API
        pagination.nextPage();
        await waitForNSubscribeCallback(products, 2);
        expect(pagination.state).toMatchSnapshot();
      });
    });
    // TODO: test with pager and sort criteria
  });
});

async function waitForNSubscribeCallback(
  controller: Controller,
  subscribeThreshold: number
) {
  await new Promise<void>((resolve) => {
    let subscribeCount = 0;
    controller.subscribe(() => {
      if (++subscribeCount < subscribeThreshold) {
        return;
      }
      resolve();
    });
  });
}
