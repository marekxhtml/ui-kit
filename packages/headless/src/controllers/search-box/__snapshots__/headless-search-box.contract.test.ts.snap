// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBox > when sorting by fields > should return search results 1`] = `
{
  "appliedSort": {
    "by": "fields",
    "fields": [],
  },
  "availableSorts": [
    {
      "by": "fields",
      "fields": [],
    },
  ],
}
`;

exports[`SearchBox > when there are more products available > should return search results 1`] = `
{
  "page": 0,
  "pageSize": 48,
  "totalEntries": 632,
  "totalPages": 14,
}
`;

exports[`SearchBox > when there are search results available > should return search results 1`] = `
[
  {
    "additionalFields": {
      "material": "cotton",
      "size": "M",
    },
    "children": [
      {
        "additionalFields": {
          "material": "cotton",
          "size": "M",
        },
        "clickUri": "https://example.com/product/0000003035-45",
        "ec_brand": "ACME",
        "ec_category": "T-Shirts",
        "ec_color": "Blue",
        "ec_description": "A very nice T-Shirt. Comes in blue and green colors.",
        "ec_gender": "M",
        "ec_images": [
          "https://example.com/image1.jpg",
        ],
        "ec_in_stock": true,
        "ec_item_group_id": "0000003035",
        "ec_listing": "T-Shirt",
        "ec_name": "ACME Corp Summer Trends",
        "ec_price": 19.99,
        "ec_product_id": "0000003035-45",
        "ec_promo_price": 14.99,
        "ec_rating": 4.5,
        "ec_shortdesc": "A very nice T-Shirt.",
        "ec_thumbnails": [
          "https://example.com/thumbnail1.jpg",
        ],
        "excerpt": "The T-Shirt is very nice",
        "excerptHighlights": [
          {
            "length": 4,
            "offset": 15,
          },
        ],
        "nameHighlights": [
          "ACME",
        ],
        "permanentid": "0000003035-45",
      },
    ],
    "clickUri": "https://example.com/product/0000003035-45",
    "ec_brand": "ACME",
    "ec_category": "T-Shirts",
    "ec_color": "Blue",
    "ec_description": "A very nice T-Shirt. Comes in blue and green colors.",
    "ec_gender": "M",
    "ec_images": [
      "https://example.com/image1.jpg",
    ],
    "ec_in_stock": true,
    "ec_item_group_id": "0000003035",
    "ec_listing": "T-Shirt",
    "ec_name": "ACME Corp Summer Trends",
    "ec_price": 19.99,
    "ec_product_id": "0000003035-45",
    "ec_promo_price": 14.99,
    "ec_rating": 4.5,
    "ec_shortdesc": "A very nice T-Shirt.",
    "ec_thumbnails": [
      "https://example.com/thumbnail1.jpg",
    ],
    "excerpt": "The T-Shirt is very nice",
    "excerptHighlights": [
      {
        "length": 4,
        "offset": 15,
      },
    ],
    "nameHighlights": [
      "ACME",
    ],
    "permanentid": "0000003035-45",
    "position": 1,
  },
]
`;

exports[`SearchBox > when there are suggestions available > should return suggestions 1`] = `
{
  "isLoading": false,
  "isLoadingSuggestions": false,
  "searchBoxId": "search_boxi",
  "suggestions": [
    {
      "highlightedValue": "suede",
      "rawValue": "suede",
    },
  ],
  "value": "",
}
`;
